File: cukcuk clone/app/src/main/java/vn/com/misa/cukcukstarterclone/ui/login/register/RegisterPresenter.java
package vn.com.misa.cukcukstarterclone.ui.login.register;

import vn.com.misa.cukcukstarterclone.data.IOnLoadedCallback;
import vn.com.misa.cukcukstarterclone.data.repository.IUserRepository;

/**
 * @created_by KhanhNQ on 08-Feb-2021.
 */
public class RegisterPresenter implements RegisterContract.Presenter {

    private RegisterContract.View view;

    private final IUserRepository userRepository;

    public RegisterPresenter(IUserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public void register(String username, String password) {
        userRepository.register(username, password, new IOnLoadedCallback<Boolean>() {
            @Override
            public void onSuccess(Boolean isSuccess) {
                if (isSuccess) {
                    view.registerSuccess();
                }
            }

            @Override
            public void onFailure(Exception e) {
                view.showErrorMessage(e.getMessage());
            }
        });
    }

    @Override
    public void attach(RegisterContract.View view) {
        this.view = view;
    }

    @Override
    public void detach() {
        view = null;
    }
}

File: app/src/main/java/com/misa/fresher/ui/login/register/RegisterPresenter.kt
package com.misa.fresher.ui.login.register

import com.misa.fresher.data.model.User
import com.misa.fresher.data.retrofit.Api
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class RegisterPresenter : RegisterContract.Presenter {
    private var view: RegisterContract.View? = null
    override fun register(user: User) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = Api.apiClient.signIn(user)
                if (response.isSuccessful && response.body() != null) {
                    withContext(Dispatchers.Main){
                        view?.registerSuccess()
                    }
                } else {
                    withContext(Dispatchers.Main) {
                        view?.showErrorMessage(response.errorBody().toString())
                    }
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    view?.showErrorMessage(e.printStackTrace().toString())
                }
            }
        }
    }

    override fun attach(view: RegisterContract.View) {
        this.view = view
    }

    override fun detach() {
        this.view = null
    }

}